### DISCUSSION TITLE
buckyos RPC design

### DISCUSSION THREAD

**Discussion by wugren** [buckyos/buckyos#50] [Ideas]
rpc接口是buckyos对外提供服务的方式，也是唯一的方式，所有的用户态程序都必须通过它来跟buckyos交互。rpc接口在系统中是名为app-server的进程，它运行于buckyos的每个节点设备上，主要工作是将该节点上的服务提供给zone内其它节点或zone外节点。

## rpc接口模块职责

1. 权限管理，支持完整的接口访问权限控制逻辑，具体的业务接口不需考虑访问权限问题。
2. 应用接口管理，app和系统服务都可以根据接口规范注册服务，注册之后就可以通过rpc调用。
3. 跨机器访问转调，让所有调用请求都仿佛本地调用。
4. 能够适应已有的x509证书应用。

## 证书需求

1. 身份证书分为根证书和子证书，根证书可以签发子证书，子证书不能再签发子证书，根证书为zone的证书，zone内子组件的证书必须有相同的根证书，子证书都由根证书签发。
2. 证书中包含版本，加密算法，组件名字、有效时间、公钥，签名、应用扩展
3. 子证书支持吊销

## rpc接口用户

rpc接口的调用用户都由用户id和组id来标识，用户id在zone之内唯一，组id包括管理员组和app组。

1. 管理员：系统进程，直接做为原生进程运行于节点设备上，系统安装时每个进程都会分配一个用户id，他们都属于管理员组。
2. zone内用户：zone内的app进程，都运行于节点设备上的隔离环境中，比如docker，app在安装时都会由app-manager分配给app一个用户id以及加入app组，并且由用户确认需要申请的权限。
3. zone外用户：跨zone访问的请求
4. 匿名用户：没有身份信息的请求

## rpc接口权限级别

1. 管理员权限接口：只有系统管理员才可以访问的接口
2. 普通权限接口：默认只有管理员才能访问，zone内用户授权之后才可以访问，zone外用户和匿名用户不能访问。
3. 无权限接口：没有权限，任何用户都可以访问

## rpc接口调用方式

1. 安全通信

   本地调用：http或https

   远程调用：https或http+其它加密协议
2. 用户信息认证

   从调用请求中获取用户身份信息，判断用户类型，是管理员、zone内app、zone外用户，如果请求中没带身份信息则为匿名用户。

   身份信息的获取方式包括多种方式，包括请求中附带身份证书、本地进程信息、静态token、用户名密码等，他们可根据需求应用于不同场景。
3. 鉴权

   判断用户是否有访问该接口的权限

   如果是管理员那直接可以访问

   如果是zone内app，那要根据安装时的授权情况判断是否可以访问

   如果是zone外用户和匿名用户则只能接口为无权限时才能访问
4. 系统进程调用过程

   a.在请求中带上用户id和组id，以及所属节点密钥签名。

   b.目标节点收到请求之后校验签名身份信息是否有效，包括签名身份证书以及请求中的签名。
5. app进程调用过程

   a.请求中不需带上身份信息，只需将访问请求发送到本节点rpc接口。

   b.rpc接口收到请求后，通过请求源接口获取到app的用户id和组id

   c.判断用户是否有访问该接口的权限，如果没有则直接返回拒绝信息

   d.判断接口服务是否本节点支持，如果支持则直接调用并返回结果。

   e.调用接口本节点不支持时，需要根据接口所属的服务名称通过nameservice查询服务的连接信息，再将请求附上本节点的签名转发到指定节点。

   f.目标节点收到请求之后判断签名是否有效，以及用户是否有访问接口的权限，只有都验证通过才调用服务接口并返回结果。

## rpc接口注册信息

1. 接口版本
2. 接口服务名词
3. 接口服务所属用户
4. 服务本地调用接口信息，可以支持unix socket、本地环路等方式。
5. 服务调用所需权限，权限包括：只允许管理员访问、只允许zone内访问、无权限访问
6. 服务提供的接口列表，每个接口包括接口名称，访问方式（比如GET、POST），访问参数

## rpc接口注册过程

rpc模块提供注册服务接口，该接口只有管理员权限才能调用。

1. 系统服务注册

   a.调用rpc模块的注册服务接口传入服务注册信息。

   b.注册服务接口验证调用者身份。

   c.发布服务到nameservice中，让其它节点可以发现。

   d.进入服务状态监控逻辑，主要判断服务所属进程是否存在，不存在则删除服务记录。
2. app服务注册

   a.app配置中定义需要注册的服务信息

   b.app-manager启动app时获取app中注册服务信息，并生成服务调用接口地址信息，比如app运行容器ip等。

   c.app-manager调用rpc模块的注册服务接口，进入系统服务注册过程。

**Comment by waterflier** [comment]
谢谢！ RPC是非常重要的，影响到所有人的基础模块。我们要尽早完成相关设计。
我已经写了一篇关于kRPC的设计 ，重点是关注从CYFS的实践中得到的一些关于鉴权的经验，可以参考看一下
https://github.com/buckyos/buckyos/blob/main/doc/PM/kRPC/proposal.md


我想kRPC的整体设计应该包含下面几个话题
0. 权限控制，有需要的内核模块应该如何利用kRPC提供的基本鉴权功能+ sys_acl提供的基本权限控制功能来实现必要的权限控制
1. 典型的调用链是怎样的，站在用户态程序来看，kRPC提供了哪些保障。站在内核服务的角度来看，如何让自己的功能可以被用户态应用基于kRPC访问到，kRPC此时又提供了哪些保障和假设？
2. 作为NetworkOS的一个经典话题，我们应该如何提高kRPC的性能（减少内存负责复制，利用好Named-Object的不可变特性做好单机缓存），这里也许需要使用一些共享内存的技巧。

**Comment by wugren** [comment]
你写的文档看过了，你的文档只是提了几种授权方式，没有说明buckyos的rpc该怎么实现，上面的方案考虑了你提的几种方式，是rpc的一种实现方式。
上面的问题1我认为在方案中是有说明的。
问题2是更具体的实现方式了，上面的方案还没有涉及。
